File: Kingdom-Fall-UI-Enabled/app/src/application/GameLogic.java
package application;

import controller.FenetreAppController;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class GameLogic extends Thread {
    private volatile boolean running = true; // Contrôle l'exécution du thread

    public Jeu jeuEnCours;
    private ExecutorService executor;
    private CountDownLatch latchFinTour;

    public GameLogic(FenetreAppController controller)
    {
        super();
        controller.setThread(this);
        executor = Executors.newSingleThreadExecutor();
        jeuEnCours = new Jeu(this, controller);
    }

    @Override
    public void run() {
        jeuEnCours.lancerJeu();

        while (running) {
            // Logique du jeu ici
            updateGame(); // Met à jour l'état du jeu
            try {
                Thread.sleep(100); // Contrôle la vitesse de mise à jour
            } catch (InterruptedException e) {
                // Gérer l'interruption du thread
                Thread.currentThread().interrupt();
            }
        }
    }

    public Jeu getJeuEnCours()
    {
        return jeuEnCours;
    }

    public void stopGame() {
        running = false; // Arrête le jeu
        this.interrupt(); // Réveille le thread si en pause
    }

    private void updateGame() {
        // Code pour mettre à jour le jeu (mouvement, collision, etc.)
    }

    public void attendreFinTour() {
        try {
            latchFinTour = new CountDownLatch(1);
            latchFinTour.await();
        }
        catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public void relacherLatch()
    {
        latchFinTour.countDown();
    }

    public void attaque() {
        executor.submit(() -> jeuEnCours.effectuerAttaque());
    }

    public void ramasser() {
        jeuEnCours.ramasser();
        relacherLatch();
    }
}

File: KF-Remake/Kingdom-Fall/source/application/GameLogic.cs
namespace App;

using System.Threading.Tasks;

public class GameLogic
{
    public Jeu JeuEnCours;
    private TaskCompletionSource<bool>? _latchFinTour;
    private readonly controllers.FenetreAppController _controller;

    public GameLogic(controllers.FenetreAppController controller)
    {
        _controller = controller;
        _controller.SetThread(this);
        JeuEnCours = new Jeu(controller, this);
    }

    public async Task Run()
    {
        await Task.Run(() => JeuEnCours.LancerJeu());
    }

    public async Task AttendreFinTour()
    {
        _latchFinTour = new TaskCompletionSource<bool>();
        await _latchFinTour.Task;
    }

    public void RelacherLatch()
    {
        if (_latchFinTour is not null && !_latchFinTour.Task.IsCompleted)
        {
            _latchFinTour?.SetResult(true);
        }
    }

    public void Attaque()
    {
        Task.Run(() => JeuEnCours.EffectuerAttaque());
    }

    public void Ramasser()
    {
        try
        {
            JeuEnCours.Ramasser();
            _controller.DissiperDrop();
        }
        catch (exceptions.InventoryFullException)
        {
            _controller.InventairePlein();
        }
    }

    public void UpdateUsername(string playerName)
    {
        JeuEnCours.UpdateUsername(playerName);
    }

    public void Equiper()
    {
        try
        {
            JeuEnCours.Equiper();
            _controller.DissiperDrop();
        }
        catch (exceptions.InventoryFullException)
        {
            _controller.InventairePlein();
        }
    }
}
